### ISSUE TITLE
mvapich-gdr error when specific MPI not specified in spack.yaml

### ISSUE THREAD

**Issue by kurtsansom** [LLNL/spack#18178]
installing from `spack.yaml` file  fails because missing `mvapich2-gdr` which is only available via binary release.

Other versions of MPI should take precedence

### Steps to reproduce the issue

create a spack.yaml file with the following content
```
spack:
  specs:
    - hdf5+hl+mpi
    - netcdf-c
    - metis
    - netlib-lapack
    - petsc+mpi
  concretization: together
```

```console
$ spack env create test spack.yaml
$ spack env activate test
$ spack install

```

### Error Message

```console
$ spack --debug --stacktrace install
build/spack/lib/spack/spack/cmd/__init__.py:122 ==> [2020-08-19-13:18:34.425411] Imported install from built-in commands
build/spack/lib/spack/spack/config.py:805 ==> [2020-08-19-13:18:34.426493] Reading config file /build/spack/var/spack/environments/test/spack.yaml
build/spack/lib/spack/spack/cmd/__init__.py:122 ==> [2020-08-19-13:18:34.428427] Imported install from built-in commands
build/spack/lib/spack/spack/concretize.py:662 ==> [2020-08-19-13:18:34.432438] [CONCRETIZATION]: Creating helper repository in /tmp/tmpu_6x0fhk
build/spack/lib/spack/spack/config.py:805 ==> [2020-08-19-13:18:34.456942] Reading config file /build/spack/etc/spack/defaults/repos.yaml
build/spack/lib/spack/spack/config.py:805 ==> [2020-08-19-13:18:34.458508] Reading config file /home/krs/.spack/repos.yaml
build/spack/lib/spack/spack/config.py:805 ==> [2020-08-19-13:18:34.535358] Reading config file /build/spack/etc/spack/defaults/packages.yaml
build/spack/lib/spack/spack/main.py:765 ==> [2020-08-19-13:18:34.591383] UnknownPackageError: Package 'mvapich2-gdr' not found.
build/spack/lib/spack/spack/error.py:55 ==> [2020-08-19-13:18:34.591525] Error: Package 'mvapich2-gdr' not found.
Traceback (most recent call last):
  File "/build/spack/lib/spack/spack/main.py", line 762, in main
    return _invoke_command(command, parser, args, unknown)
  File "/build/spack/lib/spack/spack/main.py", line 490, in _invoke_command
    return_val = command(parser, args)
  File "/build/spack/lib/spack/spack/cmd/install.py", line 277, in install
    concretized_specs = env.concretize()
  File "/build/spack/lib/spack/spack/environment.py", line 1002, in concretize
    return self._concretize_together()
  File "/build/spack/lib/spack/spack/environment.py", line 1042, in _concretize_together
    *self.user_specs
  File "/build/spack/lib/spack/spack/concretize.py", line 687, in concretize_specs_together
    concretization_root.concretize()
  File "/build/spack/lib/spack/spack/spec.py", line 2170, in concretize
    self._expand_virtual_packages(concretizer),
  File "/build/spack/lib/spack/spack/spec.py", line 2075, in _expand_virtual_packages
    candidates = concretizer.choose_virtual_or_external(spec)
  File "/build/spack/lib/spack/spack/concretize.py", line 120, in choose_virtual_or_external
    candidates = self._valid_virtuals_and_externals(spec)
  File "/build/spack/lib/spack/spack/concretize.py", line 97, in _valid_virtuals_and_externals
    externals = spec_externals(cspec)
  File "/build/spack/lib/spack/spack/package_prefs.py", line 165, in spec_externals
    names |= set(vspec.name for vspec in spec.package.virtuals_provided)
  File "/build/spack/lib/spack/spack/spec.py", line 1175, in package
    self._package = spack.repo.get(self)
  File "/build/spack/lib/spack/spack/repo.py", line 1190, in get
    return path.get(spec)
  File "/build/spack/lib/spack/spack/repo.py", line 90, in converter
    return function(self, spec_like, *args, **kwargs)
  File "/build/spack/lib/spack/spack/repo.py", line 650, in get
    return self.repo_for_pkg(spec).get(spec)
  File "/build/spack/lib/spack/spack/repo.py", line 90, in converter
    return function(self, spec_like, *args, **kwargs)
  File "/build/spack/lib/spack/spack/repo.py", line 882, in get
    raise UnknownPackageError(spec.name)
spack.repo.UnknownPackageError: Package 'mvapich2-gdr' not found..
```

### Information on your system

* **Spack:** 0.15.4
* **Python:** 3.6.9
* **Platform:** linux-ubuntu18.04-skylake

### Additional information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have searched the issues of this repo and believe this is not a duplicate
- [X] I have run the failing commands in debug mode and reported the output

**Comment by adamjstewart** [comment]
> installing from spack.yaml file fails because missing mvapich2-gdr which is only available via binary release.
>
> Other versions of MPI should take precedence

I think it's a little more complicated than that. For me, `spack spec mpi` chooses `openmpi`, which is the default. I think the reason that your build is crashing is that the last time the virtual dependency cache was created, `mvapich2-gdr` existed in your repo. Now that you are running this again, you must be on an older branch where this package does not yet exist.

Try running `spack clean -m` and see if that helps. If not, try wiping out your environment and starting over (deleting `.spack-env` and `spack.lock` should be sufficient).

**Comment by harisubramoni** [comment]
@harisubramoni - I am just tagging myself so that I can follow this.

**Comment by adamjstewart** [comment]
@harisubramoni I'm pretty sure this is a Spack virtual dependency cache issue and has nothing to do with your new `mvapich2-gdr` package, but I have been wrong in the past 😛

**Comment by harisubramoni** [comment]
Thank you, Adam. I appreciate your feedback.

If it is not too much trouble, is it possible to get myself automatically
added to all MVAPICH related tickets?

Thx,
Hari.

**Comment by adamjstewart** [comment]
@harisubramoni that's the idea behind the `maintainers` variable in a `package.py`. Right now, that process is manual. When a user opens a new issue to report a build error, the template asks them to run `spack maintainers <package>` and ping anyone who is listed as a maintainer. In the future, I've written a GitHub bot to automatically do this for PRs, although it isn't integrated yet.

**Comment by harisubramoni** [comment]
Thanks Adam! I will look at the maintainers variable.

**Comment by germasch** [comment]
FWIW, I had the same error show up after upgrading spack to 0.15.4 (from 0.14.x), and `spack clean -m` fixed it for me.

**Comment by kurtsansom** [comment]
`spack clean -m` fixed the issue. 

what could be added to the documentation and where would it go such that a users find this solution? I would be happy to make that contribution with some guidance.

**Comment by adamjstewart** [comment]
@scheibelp @becker33 I wonder if there's a way to check whether or not the virtual cache is out of date so that users don't have to run `spack clean -m`.

**Comment by kurtsansom** [comment]
I had what I think was a related issue where I updated the repo and then got weird mpi error `==> Error: There are no valid versions for mpi that match ':'` when I tried installing a new package from a separate repo with a mpi variant dependency.

calling `spack spec mpi` gave me the same error.

`spack clean -m` didn't appear to fix it, and I removed the `.spack-env` file , `spack clean -a`, and opened a new terminal and `spack spec mpi` worked and calling the same command in the specific environment also worked.
